snuba/subscriptions/consumer.py

```diff
@@ -0,0 +1,143 @@
+    # to crash and restart, it would resume at offset 2, causing the next tick
+    # interval returned to be for the messages C and D -- in this case, B and C
+    # was never returned! To avoid skipping intervals when restarting, the
+    # consumer would have had to commit the offset 1 (the offset of message B)
+    # to ensure that upon restart, the next interval would be the interval
+    # between B and C, since the message B was the first message received by
+    # the consumer.
+
+    def __init__(self, consumer: Consumer[Any]) -> None:
+        self.__consumer = consumer
+        self.__previous_messages: MutableMapping[
+            Partition, Optional[MessageDetails]
+        ] = {}
+
+    def subscribe(
+        self,
+        topics: Sequence[Topic],
+        on_assign: Optional[Callable[[Mapping[Partition, int]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[Partition]], None]] = None,
+    ) -> None:
+        def assignment_callback(partitions: Mapping[Partition, int]) -> None:
+            for partition in partitions:
+                self.__previous_messages[partition] = None
+
+            if on_assign is not None:
+                on_assign(partitions)
+
+        def revocation_callback(partitions: Sequence[Partition]) -> None:
+            # TODO: This is probably not necessary -- this could be handled as
+            # part of the assignment callback to avoid having to force reset
+            # the partitions on rebalance (see the consumer implementation.)
+            for partition in partitions:
+                del self.__previous_messages[partition]
+
+            if on_revoke is not None:
+                on_revoke(partitions)
+
+        self.__consumer.subscribe(
+            topics, on_assign=assignment_callback, on_revoke=on_revoke
+        )
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[Message[Tick]]:
+        message = self.__consumer.poll(timeout)
+        if message is None:
+            return None
+
+        previous_message = self.__previous_messages.get(message.partition)
+
+        result: Optional[Message[Tick]]
+        if previous_message is not None:
+            result = Message(
+                message.partition,
+                previous_message.offset,
+                Tick(
+                    Interval(previous_message.offset, message.offset),
+                    Interval(previous_message.timestamp, message.timestamp),
+                ),
+                previous_message.timestamp,
+            )
+        else:
+            result = None
+
+        self.__previous_messages[message.partition] = MessageDetails(
+            message.offset, message.timestamp
+        )
+
+        return result
+
+    def stage_offsets(self, offsets: Mapping[Partition, int]) -> None:
+        return self.__consumer.stage_offsets(offsets)
+
+    def commit_offsets(self) -> Mapping[Partition, int]:
+        return self.__consumer.commit_offsets()
```

Comment:
Are you sure about relying entirely on the client code to commit the right offset for this consumer knowing that they have to commit the right offset knowing they get two offsets when they poll a message ?
I think there is opportunity to shoot ourselves in the feet in a way that would be hard to troubleshoot. What about having a default commit method that automatically commit the offset that would prevent the client to consume the same interval twice ?

Reply:
> Are you sure about relying entirely on the client code to commit the right offset for this consumer knowing that they have to commit the right offset knowing they get two offsets when they poll a message ?

I think so, but I am not confident I understand this question. This doesn't differ from the implementation of any other consumer, at least when #652 is considered. If this is about changing the general offset staging/commit APIs, I think that should feedback should be given on that change (in the abstract) so it doesn't get caught up in/by the implementation details of this change. Any changes to these APIs would need to be applied at that level anyway to ensure all consumer implementations behave uniformly.

The message content (the `Tick` instance) contains an offset interval, but that doesn't really have any direct bearing on what offset is used when committing. The message offset should be used for committing, just as it would for a consumer that had any other different value type (including or not including offsets.)

The distinction between these values (`message.offset` and `message.value.offsets`) should be more apparent to some extent in the implementation of `test_tick_consumer`.

> What about having a default commit method that automatically commit the offset that would prevent the client to consume the same interval twice ?

I don't understand this, which led me to believe that I didn't really understand your earlier question/remark. Specifically:

- What do you mean by "a default commit method"?
- What do you mean by "automatically"?
- What do you mean by "the offset" (in the phrase "automatically commit the offset"?)

Reply:
Maybe this helps clarifies things, too — the message that is returned on each loop iteration actually contains the previous message consumed in that partition's offset and timestamp (e.g. if we just consumed message N, the offset of offset message N - 1 is returned):

https://github.com/getsentry/snuba/blob/6d44af5831191d015c59ac5009ed01089f3bd90b/snuba/subscriptions/consumer.py#L118-L126

This consumer will always appear to be one message behind the wrapped consumer (the one provided to the constructor) to account for this.